pallets/ctype/src/lib.rs

```diff
@@ -128,18 +140,41 @@ pub mod pallet {
 		///
 		/// # <weight>
 		/// Weight: O(1)
-		/// - Reads: Ctypes
-		/// - Writes: Ctypes
+		/// - Reads: Ctypes, Balance
+		/// - Writes: Ctypes, Balance
 		/// # </weight>
 		#[pallet::weight(<T as pallet::Config>::WeightInfo::add())]
-		pub fn add(origin: OriginFor<T>, hash: CtypeHashOf<T>) -> DispatchResult {
+		pub fn add(origin: OriginFor<T>, ctype: Vec<u8>) -> DispatchResult {
 			let source = <T as Config>::EnsureOrigin::ensure_origin(origin)?;
 			let creator = source.subject();
+			let payer = source.sender();
+
+			// Check the free balance before we do any heave work (e.g. calculate the ctype
+			// hash)
+			ensure!(
+				<T::Currency as Currency<AccountIdOf<T>>>::free_balance(&payer) >= T::Fee::get(),
```

Comment:
M: We want to use `usable_balance` here because `free_balance` includes locked amount. Please add a test case for this.

> Free balance: Non-reserved part of the balance. There may still be restrictions on this, but it is the total pool what may in principle be transferred, reserved and used for tipping.

```suggestion
				<T::Currency as Currency<AccountIdOf<T>>>::usable_balance(&payer) >= T::Fee::get(),
```

Reply:
Yes the thought is correct, but there is no `usable_balance` in the currency trait. `ensure_can_withdraw` should rather be used. Will to that